### PR TITLE
feat: support action-server within circleci pipeline (#6)

### LINKED ISSUE #6: Deploy Action Server using CircleCI
We must also connect it to the Rasa Core instances using `ACTION_SERVER_ENDPOINT`.

### PR DESCRIPTION
**Pull request by mathieubergeron**
## Description
Action server: Add support for lint/test/deploy within circleci.

image

## Related JIRA issues
closes #6 

## Checklist

- [x] PR title ends with a JIRA issue ID  
- [x] All relevant PR sections are populated, irrelevant ones are removed 
- [ ] [Draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests) is opened for WIP changes 
- [ ] PR review requested from groups not individuals 
- [ ] PR link is posted to the corresponding Slack channel

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -84,8 +115,25 @@ aliases:
...
+    parameters:
+      <<: *param__working_directory
+    working_directory: ~/project/<<parameters.working_directory>>
+    executor: python/python
+    steps:
+      - setup-project
+      - run:
+          name: Run mypy
+          command: |
+            echo test
```
Comment (dmorand) [inline .circleci/config.yml:136]: Something seems missing here 😄

File: .circleci/config.yml [hunk 2]
```diff
@@ -3,11 +3,30 @@ version: 2.1
...
+          context: org-global-v2
+      - python/black:
+          name: (AS) black
+          working_directory: action-server
+          context: org-global-v2
+      - python/test:
+          name: (AS) test
+          working_directory: action-server
+          codecov_flag: action_server
+          context: org-global-v2
```
Comment (dmorand) [inline .circleci/config.yml:24]: No isort?

File: .circleci/config.yml (hunk 1 above)
Comment (dmorand) [reply .circleci/config.yml:136]: And is it called anywhere?

File: action-server/Makefile [hunk 3]
```diff
@@ -60,6 +60,4 @@ format:
 test:
 	${ACTIVATE_VENV} py.test "${TESTS_FOLDER}/${TEST_FILE}" \
 		--cov-report term-missing:skip-covered \
-		--cov-report html \
-		--cov-fail-under=80 \
-		--cov ${SOURCE_FOLDER}
```
Comment (dmorand) [inline action-server/Makefile:65]: Why is that no longer necessary?

File: core/Dockerfile [hunk 4]
```diff
@@ -1,7 +1,7 @@
 FROM rasa/rasa:1.9.4-full
 EXPOSE 8080
 
-ENV ACTION_SERVER_ENDPOINT=http://actions:8080/webhook
+ENV ACTION_SERVER_ENDPOINT=http://action-server/webhook
```
Comment (dmorand) [inline core/Dockerfile:4]: This will no longer work using docker-compose I think...

**Comment by dmorand** [comment]
Sorry, I jumped in too early 😄

File: action-server/Makefile (hunk 3 above)
Comment (mathieubergeron) [reply action-server/Makefile:65]: those are now included in the `setup.cfg` file

File: .circleci/config.yml [hunk 5]
```diff
@@ -3,11 +3,38 @@ version: 2.1
...
+      - app/build:
+          name: (AS) build
+          component: action-server
+          context: org-global-v2
```
Comment (mathieubergeron) [inline .circleci/config.yml:36]: dès qu'on a du code python dans `core`, il va falloir ajouter toutes ces tâches aussi.

File: action-server/Makefile [hunk 6]
```diff
@@ -11,10 +11,6 @@ ACTIVATE_VENV=test -d ${VENV_LOCATION} \
 RASA_VERSION=1.9.4
 RASA_SDK_VERSION=1.9.0
 
-SOURCE_FOLDER=actions
-TESTS_FOLDER=tests
-ENV_VARS=`grep -v '^\#' .env | grep -Eo '\w+=\S+'`
```
Comment (mathieubergeron) [inline action-server/Makefile:16]: @dmorand: est-ce qu'on avait une intention avec ENV_VARS ?

File: action-server/setup.cfg [hunk 7]
```diff
@@ -0,0 +1,48 @@
+[tool:pytest]
```
Comment (mathieubergeron) [inline action-server/setup.cfg:1]: @dmorand Can you review what's in there? I've mostly took what's from beautiful-mind.

File: core/Dockerfile [hunk 8]
```diff
@@ -1,8 +1,7 @@
 FROM rasa/rasa:1.9.4-full
 EXPOSE 8080
 
-ENV ACTION_SERVER_ENDPOINT=http://actions:8080/webhook
-ENV TRACKER_STORE_ENDPOINT=tracker_store
+ENV TRACKER_STORE_ENDPOINT=tracker-store
```
Comment (fparga) [inline core/Dockerfile:4]: Is this necessary?

File: core/endpoints.yml [hunk 9]
```diff
@@ -1,5 +1,5 @@
 action_endpoint:
-  url: ${ACTION_SERVER_ENDPOINT}
+  url: http://action-server:8080/webhook
```
Comment (fparga) [inline core/endpoints.yml:2]: Couldn't we start the action-server on port `80` instead?

File: action-server/Makefile (hunk 6 above)
Comment (dmorand) [reply action-server/Makefile:16]: C'était pour charger les variables d'environnement dans le fichier `.env` mais je ne crois pas qu'on va en avoir besoin dans le contexte de ce projet. On le remettra au besoin.

File: core/Dockerfile (hunk 8 above)
Comment (mathieubergeron) [reply core/Dockerfile:4]: nop, it should be in docker-compose file actually. moving.

File: action-server/setup.cfg [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    --cov-report term
+    --cov-report term-missing
+    --cov-fail-under=0
+
+[pytest-watch]
+ext = .py, .yaml, .cfg
+
+[coverage:run]
+branch = True
+omit = tests/*,.venv/*,venv/*
```
Comment (dmorand) [inline action-server/setup.cfg:18]: `venv/*` is not necessary with out current setup (we only use `.venv`)

Comment (dmorand) [reply action-server/setup.cfg:18]: There are a few more occurences below.

File: core/endpoints.yml (hunk 9 above)
Comment (mathieubergeron) [reply core/endpoints.yml:2]: the docker container actually runs on non-root user, so we can't bind on 80.

File: action-server/setup.cfg [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+[mypy]
+mypy_path = stubs
+warn_redundant_casts = True
+warn_unused_ignores = True
+warn_return_any = True
+warn_unused_configs = True
+check_untyped_defs = True
+strict_optional = True
+namespace_packages = True
```
Comment (dmorand) [inline action-server/setup.cfg:48]: Je ne connais pas l'effet de l'ensemble de ces settings mais comme ils viennent de beautiful-mind, on pourrait les garder tels-quels jusqu'à ce que quelque chose pose problème.

BTW, @vlavoie a aussi ajouté ce fichier de config dans sa branche. Il faudrait vous synchroniser.

File: action-server/setup.cfg (hunk 10 above)
Comment (mathieubergeron) [reply action-server/setup.cfg:18]: CI/CD actually uses `venv`

File: action-server/setup.cfg (hunk 11 above)
Comment (mathieubergeron) [reply action-server/setup.cfg:48]: faudrait voir qui merge en premier ;)

**Pull request merged**